snap/quota/quota.go

```diff
@@ -257,18 +259,39 @@ func ResolveCrossReferences(grps map[string]*Group) error {
 
 // tree recursively returns all of the sub-groups of the group and the group
 // itself.
-func (grp *Group) tree() []*Group {
-	// TODO: should we be paranoid about circular references and
-	// recursion here?
-	treeList := grp.subGroups
+func (grp *Group) tree(acc []*Group) ([]*Group, error) {
+	// make sure that none of the sub-groups seen already are present in the
+	// sub-groups of this group
 	for _, sub := range grp.subGroups {
-		treeList = append(treeList, sub.tree()...)
+		// check if we have already seen this sub-group
+		for _, alreadySeen := range acc {
+			if sub == alreadySeen {
+				return nil, fmt.Errorf("internal error: circular reference found")
+			}
+		}
+
+		// check if this sub-group is actually the same group
+		if sub == grp {
```

Comment:
Nitpick^2, I think this could be the first check before the "more complex" check for already seen?

Reply:
done